hostapd: compile with OCV and provide configuration method

huaracheguarache opened the pull request:
Operating Channel Validation (OCV) is a relatively recent security feature that prevents person-in-the-middle multi-channel attacks, and it's described in detail in the following paper: https://papers.mathyvanhoef.com/wisec2018.pdf

This pull request enables the compilation of OCV for -basic and -full variants of hostapd, wpa_supplicant, and wpad. A configuration method through the wireless config file is provided, and a method to check whether the installed hostaps/wpa_supplicant/wpad version has OCV support is also provided (for future luci integration).

**I have only tested whether this works with hostapd, not with wpa_supplicant.**

Compiling with OCV support results in a small size increase (sizes calculated prior to [update to v2.10](https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=adb8c09a83d3f8e1d9e9fcbb8189b415ac0f6e86)):

**Size of wpad-basic-openssl:**
Without OCV: 731509 bytes
With OCV: 739701 bytes
Difference: 8192 bytes (1.12% relative increase)

**Size of wpad-basic-wolfssl:**
Without OCV: 727413 bytes
With OCV: 735605 bytes
Difference: 8192 bytes (1.13% relative increase)

huaracheguarache commented:
Hi, @nbd168! What do you think about this? Anything I need to change?

blocktrron commented:
FTR

```
22:06 <blocktrron> huaracheguarache: looks good as far as i can tell
22:06 <blocktrron> one question: does this feature depend on the presence of a TLS libary or can it work with a non-crypto full build?

[---]

23:36 <blocktrron> wtf, hostapd magically overrides 80211w settings in case ocv is enabled
23:38 <huaracheguarache> Yeah, that's intended behaviour per the hostapd config: # Enabling this automatically also enables ieee80211w, if not yet enabled.
23:39 <blocktrron> yes, but it is when parsing the config file
23:40 <blocktrron> and to me it looks like it depends on whether & in which order the pmf config is consumed
23:40 <blocktrron> it's not openwrts fault, but it looks off
23:41 <blocktrron> when i read the ocv config, enable pmf, read the pmf config and disable pmf
```

huaracheguarache committed:
hostapd: enable compilation of OCV and add build feature discovery

Operating Channel Validation (OCV) is a security feature designed to
prevent person-in-the-middle multi-channel attacks. Compile the -basic and
-full variants of hostapd with this feature, and enable discovery of this
feature for future luci integration. OCV can be configured by setting ocv
equal to one of the following values in the wireless config:

0 = disabled (hostapd/wpa_supplicant default)
1 = enabled
2 = enabled in workaround mode - Allow STA that claims OCV capability to
    connect even if the STA doesn't send OCI or negotiate PMF.

Signed-off-by: Michael Yartys <michael.yartys@protonmail.com>

huaracheguarache committed:
wpa_supplicant: compile with OCV support

Operating Channel Validation (OCV) is a security feature designed to
prevent person-in-the-middle multi-channel attacks. Compile -basic and
-full variants with support for OCV. This feature can be configured in the
wireless config by setting ocv equal to one of the following values:

0 = disabled (hostapd/wpa_supplicant default)
1 = enabled if wpa_supplicant's SME in use. Otherwise enabled only when the
    driver indicates support for operating channel validation.

Signed-off-by: Michael Yartys <michael.yartys@protonmail.com>

huaracheguarache commented:
@blocktrron I've done a couple of things:

* I tested whether this feature works with the full wpad without openssl/wolfssl, and it does. That's why I now compile all -basic and -full variants with OCV support.
* I noticed that I didn't actually check whether a value for `ocv` had been provided in the wireless config before adding `ocv=whatever` to the hostapd config. I've edited the commit to do that.
* I split the changes to wpa_supplicant into a separate commit. It's configured by a different part of the hostapd.sh script, and I failed to notice that I didn't provide a method to configure it. It's now possible to do so.
* I tested whether the order of `ocv` relative to `ieee80211w` in the hostapd config matters, and it does. What happens is that if `ieee80211w` is configured before `ocv` hostapd will silently override the PMF setting in case it's disabled. If it's the other way around and PMF is disabled hostapd will fail to start. See an example of this here:

```
Sun Feb 13 16:21:08 2022 daemon.err hostapd: OCV: PMF needs to be enabled whenever using OCV
Sun Feb 13 16:21:08 2022 daemon.err hostapd: 1 errors found in configuration file '/var/run/hostapd-phy1.conf'
Sun Feb 13 16:21:08 2022 daemon.err hostapd: Failed to set up interface with /var/run/hostapd-phy1.conf
Sun Feb 13 16:21:08 2022 daemon.notice netifd: radio1 (3820): Command failed: Invalid argument
Sun Feb 13 16:21:08 2022 daemon.notice netifd: radio1 (3820): Device setup failed: HOSTAPD_START_FAILED
Sun Feb 13 16:21:08 2022 daemon.notice netifd: Wireless device 'radio1' set retry=0
Sun Feb 13 16:21:08 2022 daemon.crit netifd: Wireless device 'radio1' setup failed, retry=0
Sun Feb 13 16:21:08 2022 daemon.notice netifd: Wireless device 'radio1' is now down
```
My opinion is that it's preferable for hostapd to fail to start and complain in the log instead of silently changing settings. That's why I've placed the code that sets `ocv` before the code that sets PMF in the hostapd.sh script.

huaracheguarache commented:
@blocktrron So, currently the hostapd part works well, but I've discovered that wpa_supplicant doesn't automatically set ieee80211w=1 when setting ocv=1, even though it's supposed to do this. I guess it's possible to just merge the part that changes the hostapd stuff, but ideally this bug would get fixed in upstream first so that the entire PR can be merged.

hauke commented:
Did you report this problem to upstream? If so, please share a link to the mailing list post.

huaracheguarache commented:
@hauke Here's my post to the hostapd mailing list: http://lists.infradead.org/pipermail/hostap/2022-February/040239.html

aparcar commented:
Looks good to me, @hauke do you think we can proceed?

tiagogaspar8 commented:
Any news about this?

huaracheguarache commented:
After a long hiatus I've rebased the code on top of master. Would appreciate if anyone else could help with testing this PR to confirm whether things work as they should.

hauke commented:
Thank you for your patch I applied them to master in commit f60628f33ca9891fd9fb814588530df56ebdcdca and 442708dfe2f599796b87eb113b03ab6cc42e0292.

silentcreek commented:
Could this please be backported to 22.03? I've noticed on the Wiki that ocv is now available and was eager to try it on the new stable release, only to realize later that this is not yet available in 22.03, unfortunately.
